# Name- Sangramjit Hazarika
# Registration Number- 19BCE1371
# Lab-12

### Q) Take a collection of documents. Implement automatic topic extraction using LDA.  Draw your inferences.

## Import Libraries

In [29]:
from nltk.corpus import stopwords 
from nltk.stem.wordnet import WordNetLemmatizer
import string

# Import numpy for matrix operation
import numpy as np

# Importing Gensim
import gensim
from gensim import corpora

In [30]:
from pathlib import Path
doc1 = Path('/Users/sangramjithazarika/Web Mining Lab 5th Sem/19BCE1371_Lab12_WebMining_CSE3004/doc1.txt').read_text()
doc2 = Path('/Users/sangramjithazarika/Web Mining Lab 5th Sem/19BCE1371_Lab12_WebMining_CSE3004/doc2.txt').read_text()
doc3 = Path('/Users/sangramjithazarika/Web Mining Lab 5th Sem/19BCE1371_Lab12_WebMining_CSE3004/doc3.txt').read_text()
doc4 = Path('/Users/sangramjithazarika/Web Mining Lab 5th Sem/19BCE1371_Lab12_WebMining_CSE3004/doc4.txt').read_text()
doc5 = Path('/Users/sangramjithazarika/Web Mining Lab 5th Sem/19BCE1371_Lab12_WebMining_CSE3004/doc5.txt').read_text()
doc6 = Path('/Users/sangramjithazarika/Web Mining Lab 5th Sem/19BCE1371_Lab12_WebMining_CSE3004/doc6.txt').read_text()
doc7 = Path('/Users/sangramjithazarika/Web Mining Lab 5th Sem/19BCE1371_Lab12_WebMining_CSE3004/doc7.txt').read_text()
doc8 = Path('/Users/sangramjithazarika/Web Mining Lab 5th Sem/19BCE1371_Lab12_WebMining_CSE3004/doc8.txt').read_text()
doc9 = Path('/Users/sangramjithazarika/Web Mining Lab 5th Sem/19BCE1371_Lab12_WebMining_CSE3004/doc9.txt').read_text()
doc10 = Path('/Users/sangramjithazarika/Web Mining Lab 5th Sem/19BCE1371_Lab12_WebMining_CSE3004/doc10.txt').read_text()


In [31]:
# combining all the documents into a list:

corpus = [doc1, doc2, doc3, doc4, doc5, doc6, doc7, doc8, doc9, doc10]
print(corpus)

['Russian President Vladimir Putin says a list of officials and businessmen close to the Kremlin published by the US has in effect targeted all Russian people.\n\nThe list names 210 top Russians as part of a sanctions law aimed at punishing Moscow for meddling in the US election.\n\nHowever, the US stressed those named were not subject to new sanctions.\n\nMr Putin said the list was an unfriendly act that complicated US-Russia ties but he said he did not want to escalate the situation.\n\nMr Putin said Russia should instead be thinking about ""ourselves and the economy"".\n\nThe list was also derided by a number of senior Russian officials who said it bore a strong resemblance to the Forbes magazine ranking of Russian billionaires. A US Treasury Department later told Buzzfeed that an unclassified annex of the report had been derived from the magazine.', 'Why did the US publish the list?\n\nThe government was required to draw up the list after Congress passed the Countering America\'s A

In [32]:
# The complete set of documents (corpus)

corpus

['Russian President Vladimir Putin says a list of officials and businessmen close to the Kremlin published by the US has in effect targeted all Russian people.\n\nThe list names 210 top Russians as part of a sanctions law aimed at punishing Moscow for meddling in the US election.\n\nHowever, the US stressed those named were not subject to new sanctions.\n\nMr Putin said the list was an unfriendly act that complicated US-Russia ties but he said he did not want to escalate the situation.\n\nMr Putin said Russia should instead be thinking about ""ourselves and the economy"".\n\nThe list was also derided by a number of senior Russian officials who said it bore a strong resemblance to the Forbes magazine ranking of Russian billionaires. A US Treasury Department later told Buzzfeed that an unclassified annex of the report had been derived from the magazine.',
 'Why did the US publish the list?\n\nThe government was required to draw up the list after Congress passed the Countering America\'s 

## Text Preprocessing

In [33]:
# Apply Preprocessing on the Corpus

# Stop Words Removal
stop = set(stopwords.words('english'))

# Removal of Punctuation Marks
exclude = set(string.punctuation) 

# Lemmatization
lemma = WordNetLemmatizer() 

def clean(doc):
    
    # Convert text into lower case + split into words
    stop_free = " ".join([i for i in doc.lower().split() if i not in stop])
    
    # Remove any stop words present
    punc_free = ''.join(ch for ch in stop_free if ch not in exclude)  
    
    # Remove punctuations + normalize the text
    normalized = " ".join(lemma.lemmatize(word) for word in punc_free.split())  
    return normalized

# Clean data stored in a new list
clean_corpus = [clean(doc).split() for doc in corpus]  

In [34]:
clean_corpus

[['russian',
  'president',
  'vladimir',
  'putin',
  'say',
  'list',
  'official',
  'businessmen',
  'close',
  'kremlin',
  'published',
  'u',
  'effect',
  'targeted',
  'russian',
  'people',
  'list',
  'name',
  '210',
  'top',
  'russian',
  'part',
  'sanction',
  'law',
  'aimed',
  'punishing',
  'moscow',
  'meddling',
  'u',
  'election',
  'however',
  'u',
  'stressed',
  'named',
  'subject',
  'new',
  'sanction',
  'mr',
  'putin',
  'said',
  'list',
  'unfriendly',
  'act',
  'complicated',
  'usrussia',
  'tie',
  'said',
  'want',
  'escalate',
  'situation',
  'mr',
  'putin',
  'said',
  'russia',
  'instead',
  'thinking',
  'ourselves',
  'economy',
  'list',
  'also',
  'derided',
  'number',
  'senior',
  'russian',
  'official',
  'said',
  'bore',
  'strong',
  'resemblance',
  'forbes',
  'magazine',
  'ranking',
  'russian',
  'billionaire',
  'u',
  'treasury',
  'department',
  'later',
  'told',
  'buzzfeed',
  'unclassified',
  'annex',
  'report'

## Creating Document Term Matrix

The advantage of using gensim for Document Term Matrix(DTM) is that we don't need to create the matrix from scratch explicitly. The gensim library has internal mechanism to create the Document Term Matrix(DTM).

The only requirement for gensis package is we need to pass the cleaned data in the form of tokenized words.

In [35]:
# Creating the dictionary containing all the terms of our courpus(Specific to Genism syntax perspective)

dict_ = corpora.Dictionary(clean_corpus)

print(dict_)

Dictionary(734 unique tokens: ['210', 'act', 'aimed', 'also', 'annex']...)


In [36]:
# Printing the terms
for i in dict_.values():
    print(i)

210
act
aimed
also
annex
billionaire
bore
businessmen
buzzfeed
close
complicated
department
derided
derived
economy
effect
election
escalate
forbes
however
instead
kremlin
later
law
list
magazine
meddling
moscow
mr
name
named
new
number
official
ourselves
part
people
president
published
punishing
putin
ranking
report
resemblance
russia
russian
said
sanction
say
senior
situation
stressed
strong
subject
targeted
thinking
tie
told
top
treasury
u
unclassified
unfriendly
usrussia
vladimir
want
10
114
1bn
2016
96
accused
action
administration
adversary
affair
again
alexander
alleged
ally
although
america
apparently
association
august
benefited
bortnikov
caatsa
caption
chief
committee
congress
coolness
copyright
could
countering
delivered
democrat
determined
donald
draw
eliot
engel
even
fact
federal
foreign
fsb
future
government
guy
hook
house
image
include
informally
intelligence
it
key
known
letting
linked
longstanding
many
may
midnight
minute
monday
naryshkin
notice
oligarch
opposed
opposi

Now, the next step is to convert the corpus (the list of documents) into a document-term Matrix using the dictionary containing all the terms.

In [37]:
# Converting list of documents (corpus) into Document Term Matrix using the dictionary 
doc_term_matrix = [dict_.doc2bow(i) for i in clean_corpus]
doc_term_matrix

[[(0, 1),
  (1, 1),
  (2, 1),
  (3, 1),
  (4, 1),
  (5, 1),
  (6, 1),
  (7, 1),
  (8, 1),
  (9, 1),
  (10, 1),
  (11, 1),
  (12, 1),
  (13, 1),
  (14, 1),
  (15, 1),
  (16, 1),
  (17, 1),
  (18, 1),
  (19, 1),
  (20, 1),
  (21, 1),
  (22, 1),
  (23, 1),
  (24, 4),
  (25, 2),
  (26, 1),
  (27, 1),
  (28, 2),
  (29, 1),
  (30, 1),
  (31, 1),
  (32, 1),
  (33, 2),
  (34, 1),
  (35, 1),
  (36, 1),
  (37, 1),
  (38, 1),
  (39, 1),
  (40, 3),
  (41, 1),
  (42, 1),
  (43, 1),
  (44, 1),
  (45, 5),
  (46, 4),
  (47, 2),
  (48, 1),
  (49, 1),
  (50, 1),
  (51, 1),
  (52, 1),
  (53, 1),
  (54, 1),
  (55, 1),
  (56, 1),
  (57, 1),
  (58, 1),
  (59, 1),
  (60, 4),
  (61, 1),
  (62, 1),
  (63, 1),
  (64, 1),
  (65, 1)],
 [(0, 1),
  (1, 1),
  (2, 1),
  (7, 1),
  (9, 2),
  (16, 1),
  (21, 1),
  (23, 4),
  (24, 5),
  (26, 1),
  (28, 3),
  (29, 2),
  (30, 2),
  (37, 3),
  (39, 1),
  (40, 4),
  (44, 2),
  (45, 1),
  (47, 4),
  (54, 1),
  (56, 1),
  (58, 1),
  (60, 2),
  (61, 1),
  (66, 1),
  (67, 1),
  

The output implies:

    - Document wise we have the index of the word and its frequency.
    - The 0th word is repeated 1 time, then the 1st word repeated 1 and so on ...

## Implementation of LDA

In [38]:
# Creating the object for LDA model using gensim library

Lda = gensim.models.ldamodel.LdaModel

In [39]:
# Running and Training LDA model on the document term matrix.

ldamodel = Lda(doc_term_matrix, num_topics=6, id2word = dict_, passes=1, random_state=0, eval_every=None)

In [40]:
# Print the topics with the indexes: 0,1,2, etc.

ldamodel.print_topics()

[(0,
  '0.008*"russian" + 0.008*"restaurant" + 0.008*"star" + 0.007*"u" + 0.006*"said" + 0.006*"guide" + 0.006*"list" + 0.005*"michelin" + 0.005*"chef" + 0.005*"mr"'),
 (1,
  '0.016*"u" + 0.015*"list" + 0.014*"sanction" + 0.012*"said" + 0.012*"russian" + 0.011*"mr" + 0.009*"putin" + 0.008*"law" + 0.008*"russia" + 0.006*"named"'),
 (2,
  '0.013*"parker" + 0.011*"said" + 0.009*"list" + 0.008*"mr" + 0.008*"sanction" + 0.007*"u" + 0.007*"image" + 0.006*"russian" + 0.006*"putin" + 0.005*"footage"'),
 (3,
  '0.013*"said" + 0.012*"parker" + 0.008*"sanction" + 0.007*"list" + 0.007*"u" + 0.006*"restaurant" + 0.006*"top" + 0.006*"mr" + 0.006*"court" + 0.005*"russian"'),
 (4,
  '0.023*"restaurant" + 0.014*"star" + 0.008*"guide" + 0.008*"michelin" + 0.007*"food" + 0.007*"bra" + 0.007*"mr" + 0.007*"said" + 0.006*"one" + 0.006*"parker"'),
 (5,
  '0.010*"manchester" + 0.010*"parker" + 0.008*"it" + 0.007*"judge" + 0.007*"could" + 0.007*"list" + 0.007*"mr" + 0.007*"image" + 0.006*"told" + 0.006*"people

The output shows the dominance of a particular topic for the corresponding word.

## Extracting Topics from the Corpus

In [41]:
print(ldamodel.print_topics(num_topics=6, num_words=5))

# num_topics mean: how many topics want to extract 
# num_words: the number of words that want per topic

[(0, '0.008*"russian" + 0.008*"restaurant" + 0.008*"star" + 0.007*"u" + 0.006*"said"'), (1, '0.016*"u" + 0.015*"list" + 0.014*"sanction" + 0.012*"said" + 0.012*"russian"'), (2, '0.013*"parker" + 0.011*"said" + 0.009*"list" + 0.008*"mr" + 0.008*"sanction"'), (3, '0.013*"said" + 0.012*"parker" + 0.008*"sanction" + 0.007*"list" + 0.007*"u"'), (4, '0.023*"restaurant" + 0.014*"star" + 0.008*"guide" + 0.008*"michelin" + 0.007*"food"'), (5, '0.010*"manchester" + 0.010*"parker" + 0.008*"it" + 0.007*"judge" + 0.007*"could"')]


## Assigning the topics to the documents

#### Pass the doc_term_matrix to the LDA model:

In [42]:
# Printing the topic associations with the documents
count = 0
for i in ldamodel[doc_term_matrix]:
    print("Document: ",count+1,i)
    count += 1

Document:  1 [(1, 0.99026805)]
Document:  2 [(1, 0.9948704)]
Document:  3 [(3, 0.9937554)]
Document:  4 [(1, 0.8934651), (5, 0.098665394)]
Document:  5 [(1, 0.9935173)]
Document:  6 [(5, 0.99314517)]
Document:  7 [(3, 0.9939786)]
Document:  8 [(3, 0.16803543), (4, 0.71171635), (5, 0.117395744)]
Document:  9 [(4, 0.41250852), (5, 0.5812817)]
Document:  10 [(4, 0.9950594)]


## Inference

The ten documents are assigned the topics with the higher weightage for the respective document.
From above can see:

    - Document 1 resembles with Topic 1 with 99% weightage.
    - Document 2 again resembles with Topic 1 with 99.48% weightage.
    - Document 3 resembles with Topic 3 with 99.37% weightage.
    - Topic 1 dominates the document 4 with the weightage of 89.3%. 
    - Document 5 is influenced by the Topic 1 with 99.3%.
    - Topic 5 dominates Document 6 with 99.3% weightage.
    - Document 7 resembles with Topic 3 with 99.39% weightage.
    - Topic 4 dominates Document 8 with 71.2% weightage when compared to Topic 3 and 5 with 16.8% and 11.7% 
      weightage respectively.
    - Topic 5 dominates Document 9 with 58.1% weightage when compared to Topic 4 with 41.2% weightage.
    - Document 10 resembles with Topic 4 with 99%.5 weightage.

### -----------------------------------------------------Thank you Sir------------------------------------------------------------